File: src/utils/debtSimplifier.ts
import { Expense, GroupMember, DirectDebt, DebtSimplificationResult } from '@/types';

/**
 * Calculates net balance per user and applies a Greedy Min-Flow Graph Algorithm
 * to simplify N debts down to the minimal possible direct transactions.
 */
export function calculateSimplifiedDebts(
  cohortId: string,
  members: GroupMember[],
  expenses: Expense[]
): DebtSimplificationResult {
  const netBalances: Record<string, number> = {};

  // Initialize all members with 0 balance
  members.forEach((m) => {
    netBalances[m.userId] = 0;
  });

  // Calculate net balances for each expense
  expenses.forEach((expense) => {
    if (expense.cohortId !== cohortId) return;

    const paidBy = expense.paidByUserId;
    const totalAmount = expense.totalAmount;

    // Credit the person who paid
    netBalances[paidBy] = (netBalances[paidBy] || 0) + totalAmount;

    // Debit each participant according to their split amount
    expense.splits.forEach((split) => {
      netBalances[split.userId] = (netBalances[split.userId] || 0) - split.amount;
    });
  });

  // Separate debtors (people who owe money) and creditors (people owed money)
  const debtors: { userId: string; amount: number }[] = [];
  const creditors: { userId: string; amount: number }[] = [];

  Object.entries(netBalances).forEach(([userId, balance]) => {
    // Round to 2 decimal places to avoid floating point precision issues
    const rounded = Math.round(balance * 100) / 100;
    netBalances[userId] = rounded;

    if (rounded < -0.01) {
      debtors.push({ userId, amount: Math.abs(rounded) });
    } else if (rounded > 0.01) {
      creditors.push({ userId, amount: rounded });
    }
  });

  // Sort debtors and creditors descending by amount
  debtors.sort((a, b) => b.amount - a.amount);
  creditors.sort((a, b) => b.amount - a.amount);

  const simplifiedDebts: DirectDebt[] = [];

  let i = 0; // debtor index
  let j = 0; // creditor index

  // Greedy matching: match largest debtor with largest creditor
  while (i < debtors.length && j < creditors.length) {
    const debtor = debtors[i];
    const creditor = creditors[j];

    const settledAmount = Math.min(debtor.amount, creditor.amount);

    if (settledAmount > 0.01) {
      const debtorMember = members.find((m) => m.userId === debtor.userId);
      const creditorMember = members.find((m) => m.userId === creditor.userId);

      simplifiedDebts.push({
        fromUserId: debtor.userId,
        fromProfile: debtorMember?.profile,
        toUserId: creditor.userId,
        toProfile: creditorMember?.profile,
        amount: Math.round(settledAmount * 100) / 100,
        currency: expenses[0]?.currency || 'INR',
      });
    }

    debtor.amount -= settledAmount;
    creditor.amount -= settledAmount;

    if (debtor.amount < 0.01) i++;
    if (creditor.amount < 0.01) j++;
  }

  return {
    cohortId,
    netBalances,
    simplifiedDebts,
  };
}

File: lib/core/algorithms/debt_simplifier.dart
import '../models/direct_debt.dart';
import '../models/expense.dart';
import '../models/group_member.dart';
import '../models/split.dart';

class _DebtorCreditorNode {
  final String userId;
  double amount;

  _DebtorCreditorNode({
    required this.userId,
    required this.amount,
  });
}

/// DebtSimplifier implements a Greedy Min-Flow Graph algorithm to resolve
/// cyclic multi-person IOUs and reduce N-person group debts down to the
/// mathematical minimum number of direct transactions.
class DebtSimplifier {
  /// Calculates net balances per user and generates the simplified debt transactions.
  static DebtSimplificationResult calculateSimplifiedDebts({
    required String cohortId,
    required List<GroupMember> members,
    required List<Expense> expenses,
  }) {
    final Map<String, double> netBalances = {};

    // 1. Initialize all members with 0.0 balance
    for (final m in members) {
      netBalances[m.userId] = 0.0;
    }

    // 2. Calculate net balances for each expense in this cohort
    for (final expense in expenses) {
      if (expense.cohortId != cohortId) continue;

      final paidBy = expense.paidByUserId;
      final totalAmount = expense.totalAmount;

      // Credit the payer(s)
      if (expense.payers.isNotEmpty) {
        for (final payer in expense.payers) {
          netBalances[payer.userId] =
              (netBalances[payer.userId] ?? 0.0) + payer.amount;
        }
      } else {
        netBalances[paidBy] = (netBalances[paidBy] ?? 0.0) + totalAmount;
      }

      // Debit participants according to split amounts
      final splits = expense.splits.isNotEmpty
          ? expense.splits
          : [ExpenseSplit(userId: paidBy, amount: totalAmount)];

      for (final split in splits) {
        netBalances[split.userId] =
            (netBalances[split.userId] ?? 0.0) - split.amount;
      }
    }

    // 3. Separate debtors (< -0.01) and creditors (> 0.01)
    final List<_DebtorCreditorNode> debtors = [];
    final List<_DebtorCreditorNode> creditors = [];

    final Map<String, double> roundedNetBalances = {};
    for (final entry in netBalances.entries) {
      // Round to 2 decimal places to avoid floating point precision artifacts
      final rounded = (entry.value * 100).round() / 100.0;
      roundedNetBalances[entry.key] = rounded;

      if (rounded < -0.01) {
        debtors.add(_DebtorCreditorNode(
          userId: entry.key,
          amount: rounded.abs(),
        ));
      } else if (rounded > 0.01) {
        creditors.add(_DebtorCreditorNode(
          userId: entry.key,
          amount: rounded,
        ));
      }
    }

    // 4. Sort debtors and creditors descending by amount
    debtors.sort((a, b) => b.amount.compareTo(a.amount));
    creditors.sort((a, b) => b.amount.compareTo(a.amount));

    // 5. Greedy matching: match largest debtor with largest creditor
    final List<DirectDebt> simplifiedDebts = [];
    int i = 0; // debtor pointer
    int j = 0; // creditor pointer

    final currency = expenses.isNotEmpty ? expenses.first.currency : 'INR';

    while (i < debtors.length && j < creditors.length) {
      final debtor = debtors[i];
      final creditor = creditors[j];

      final settledAmount = debtor.amount < creditor.amount
          ? debtor.amount
          : creditor.amount;

      if (settledAmount > 0.01) {
        GroupMember? debtorMember;
        GroupMember? creditorMember;

        for (final m in members) {
          if (m.userId == debtor.userId) debtorMember = m;
          if (m.userId == creditor.userId) creditorMember = m;
        }

        simplifiedDebts.add(DirectDebt(
          fromUserId: debtor.userId,
          fromProfile: debtorMember?.profile,
          toUserId: creditor.userId,
          toProfile: creditorMember?.profile,
          amount: (settledAmount * 100).round() / 100.0,
          currency: currency,
        ));
      }

      debtor.amount -= settledAmount;
      creditor.amount -= settledAmount;

      if (debtor.amount < 0.01) i++;
      if (creditor.amount < 0.01) j++;
    }

    return DebtSimplificationResult(
      cohortId: cohortId,
      netBalances: roundedNetBalances,
      simplifiedDebts: simplifiedDebts,
    );
  }
}
